cirq-core/cirq/transformers/merge_single_qubit_gates.py

```diff
@@ -0,0 +1,145 @@
+from cirq.transformers import transformer_api, transformer_primitives
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@transformer_api.transformer
+def merge_single_qubit_gates(
+    circuit: 'cirq.AbstractCircuit',
+    *,
+    context: Optional['cirq.TransformerContext'] = None,
+    rewriter: Optional[Callable[[List[ops.Operation]], ops.OP_TREE]] = None,
+    synthesizer: Optional[Callable[[ops.Qid, np.ndarray], ops.OP_TREE]] = None,
+) -> 'cirq.Circuit':
+    """Merges adjacent single qubit unitaries in the given circuit.
+
+    Uses rewriter or synthesiser to convert a connected component of single qubit unitaries
+    into a more desirable form. At-most one of rewriter or synthesiser should be specified
+    at a time. If both are not specified, connected components are replaced by a single
+    `cirq.MatrixGate` containing unitary of the merged component.
+
+    Args:
+        circuit: Input circuit to transform. It will not be modified.
+        context: `cirq.TransformerContext` storing common configurable options for transformers.
+        rewriter: Specifies how to merge runs of single-qubit operations into a more desirable
+            form. Takes a list of operations and produces a list of operations.
+        synthesizer: A special kind of rewriter that operates purely on the unitary matrix of the
+            intended operation. Takes a qubit and a unitary matrix and returns a list of operations.
+
+    Returns:
+        Copy of the transformed input circuit.
+
+    Raises:
+        ValueError: If both rewriter and synthesizer were specified.
+    """
+    if rewriter is not None and synthesizer is not None:
+        raise ValueError("Can't specify both rewriter and synthesizer.")
+    merged_circuit_op_tag = "_merged_single_qubit_gates_component"
+
+    def map_func(op: 'cirq.Operation', _) -> 'cirq.OP_TREE':
+        if not (protocols.num_qubits(op) == 1 and protocols.has_unitary(op)):
+            return op
+        op_untagged = op.untagged
+        if rewriter:
+            return rewriter(
+                [*op_untagged.circuit.all_operations()]
+                if isinstance(op_untagged, circuits.CircuitOperation)
+                and merged_circuit_op_tag in op.tags
+                else [op]
+            )
+        elif synthesizer:
+            return synthesizer(op.qubits[0], protocols.unitary(op))
+        else:
+            return ops.MatrixGate(protocols.unitary(op)).on(*op.qubits)
+
+    circuit = transformer_primitives.merge_k_qubit_unitaries_to_circuit_op(
+        circuit,
+        k=1,
+        tags_to_ignore=context.tags_to_ignore if context else (),
+        merged_circuit_op_tag=merged_circuit_op_tag,
+    )
+    return transformer_primitives.map_operations_and_unroll(
+        circuit, map_func, tags_to_ignore=context.tags_to_ignore if context else ()
+    ).unfreeze(copy=False)
+
+
+@transformer_api.transformer
+def merge_single_qubit_gates_to_phased_x_and_z(
```

Comment:
I think we should still keep this transformer as it's hard to predict all the user call sites where it might be getting used and whether phsx version would be an ideal replacement or not. There's also very little overhead to keep it